File: src/Packs/FamilyOS/KitPlatform.Packs.FamilyOs.Application/FamilyPlaybookIds.cs
namespace KitPlatform.Packs.FamilyOs;

/// <summary>Wave B starter playbook IDs — stable codes for CoachInsight / AFE (not a CMS).</summary>
public static class FamilyPlaybookIds
{
    public const string BrushForgot = "PB0001";
    public const string PackForgot = "PB0002";
    public const string SleepLate = "PB0003";
    public const string SkipReading = "PB0004";
    public const string TooManyNudges = "PB0005";
    public const string EveningOverdue = "PB0006";
    public const string AwaitingCheck = "PB0007";
    public const string PatternForgot = "PB0008";
    public const string SparseBlueprint = "PB0020";
    public const string SkipForgot = "PB0017";

    public static string? Resolve(
        string? proposalCode,
        string? focusTitle,
        int patternForgotCount,
        bool blueprintSparse = false)
    {
        var title = focusTitle ?? "";
        var brush = LooksBrush(title);
        var pack = LooksPack(title);
        var sleep = LooksSleep(title);
        var read = LooksRead(title);

        if (patternForgotCount >= 3)
        {
            if (brush) return BrushForgot;
            if (pack) return PackForgot;
            if (sleep) return SleepLate;
            if (read) return SkipReading;
            return PatternForgot;
        }

        if (string.Equals(proposalCode, FamilyCoachProposalCodes.SuggestMoveAfterDinner, StringComparison.OrdinalIgnoreCase)
            || string.Equals(proposalCode, FamilyCoachProposalCodes.SuggestMoveAfterSchool, StringComparison.OrdinalIgnoreCase))
        {
            if (brush) return BrushForgot;
            if (pack) return PackForgot;
            return PatternForgot;
        }

        if (string.Equals(proposalCode, FamilyCoachProposalCodes.SupportOverdue, StringComparison.OrdinalIgnoreCase))
            return EveningOverdue;

        if (brush) return BrushForgot;
        if (pack) return PackForgot;
        if (sleep) return SleepLate;
        if (read) return SkipReading;

        // No habit/pattern signal — sparse Blueprint asks for DNA first.
        if (blueprintSparse) return SparseBlueprint;

        return null;
    }

    private static bool LooksBrush(string title)
    {
        var t = title.ToLowerInvariant();
        return t.Contains("đánh răng", StringComparison.Ordinal)
            || t.Contains("danh rang", StringComparison.Ordinal)
            || t.Contains("brush", StringComparison.Ordinal);
    }

    private static bool LooksPack(string title)
    {
        var t = title.ToLowerInvariant();
        return t.Contains("cặp", StringComparison.Ordinal)
            || t.Contains("balo", StringComparison.Ordinal)
            || t.Contains("chuẩn bị", StringComparison.Ordinal)
            || t.Contains("chuan bi", StringComparison.Ordinal)
            || t.Contains("đồng phục", StringComparison.Ordinal);
    }

    private static bool LooksSleep(string title)
    {
        var t = title.ToLowerInvariant();
        return t.Contains("ngủ", StringComparison.Ordinal) || t.Contains("ngu ", StringComparison.Ordinal);
    }

    private static bool LooksRead(string title)
    {
        var t = title.ToLowerInvariant();
        return t.Contains("đọc", StringComparison.Ordinal) || t.Contains("sách", StringComparison.Ordinal);
    }
}

File: client/family-app/src/shared/value/family-playbook-ids.ts
/**
 * Famixa Playbook starter IDs — Wave B stub (not a CMS).
 * Map gradually onto CoachInsight proposalCode / AFE proposals.
 */
export const FAMILY_PLAYBOOK_STARTER = [
  { id: 'PB0001', trigger: 'Con quên đánh răng', domain: 'habit_brush' },
  { id: 'PB0002', trigger: 'Quên chuẩn bị cặp sáng', domain: 'habit_pack' },
  { id: 'PB0003', trigger: 'Ngủ muộn lặp', domain: 'sleep' },
  { id: 'PB0004', trigger: 'Bỏ qua đọc sách', domain: 'learning' },
  { id: 'PB0005', trigger: 'Quá nhiều lần nhắc phụ huynh', domain: 'peace' },
  { id: 'PB0006', trigger: 'Việc quá giờ buổi tối', domain: 'twin_evening' },
  { id: 'PB0007', trigger: 'Con báo xong — chờ duyệt', domain: 'awaiting' },
  { id: 'PB0008', trigger: 'Pattern quên ≥3/7 ngày', domain: 'coach_insight' },
  { id: 'PB0009', trigger: 'Stage teen — tránh giọng trẻ nhỏ', domain: 'stage' },
  { id: 'PB0010', trigger: 'Values Reading → ưu tiên đọc', domain: 'values' },
  { id: 'PB0011', trigger: 'Focus tự lập / ít nhắc', domain: 'goals' },
  { id: 'PB0012', trigger: 'School bubble competitive', domain: 'context' },
  { id: 'PB0013', trigger: 'Self-view lệch peer shock', domain: 'calibration' },
  { id: 'PB0014', trigger: 'Growth Balance — tránh dễ hư', domain: 'growth_balance' },
  { id: 'PB0015', trigger: 'Growth Balance — thiếu phấn đấu', domain: 'growth_balance' },
  { id: 'PB0016', trigger: 'Movie Night / team incomplete', domain: 'team' },
  { id: 'PB0017', trigger: 'Skip lý do forgot', domain: 'skip' },
  { id: 'PB0018', trigger: 'Optional task done đúng giờ', domain: 'strength' },
  { id: 'PB0019', trigger: 'Illusion risk hit', domain: 'illusion' },
  { id: 'PB0020', trigger: 'Chưa hydrate Blueprint', domain: 'sparse' },
] as const;

export type FamilyPlaybookId = (typeof FAMILY_PLAYBOOK_STARTER)[number]['id'];

const PROPOSAL = {
  suggest_move_after_dinner: 'suggest_move_after_dinner',
  suggest_move_after_school: 'suggest_move_after_school',
  open_today: 'open_today',
  support_overdue: 'support_overdue',
} as const;

function looksBrush(title?: string | null): boolean {
  const t = (title ?? '').toLowerCase();
  return t.includes('đánh răng') || t.includes('danh rang') || t.includes('brush');
}

function looksPack(title?: string | null): boolean {
  const t = (title ?? '').toLowerCase();
  return (
    t.includes('cặp') ||
    t.includes('balo') ||
    t.includes('chuẩn bị') ||
    t.includes('chuan bi') ||
    t.includes('đồng phục')
  );
}

function looksSleep(title?: string | null): boolean {
  const t = (title ?? '').toLowerCase();
  return t.includes('ngủ') || t.includes('ngu ');
}

function looksRead(title?: string | null): boolean {
  const t = (title ?? '').toLowerCase();
  return t.includes('đọc') || t.includes('sách');
}

/**
 * Resolve stable Playbook ID from CoachInsight / Brief signals.
 * Habit/pattern wins over sparse; sparse (PB0020) only when no other signal.
 */
export function resolvePlaybookId(input: {
  blueprintSparse?: boolean;
  proposalCode?: string | null;
  focusTitle?: string | null;
  patternForgotCount?: number;
  skipReasonForgot?: boolean;
  awaitingCheck?: boolean;
}): FamilyPlaybookId | null {
  if (input.awaitingCheck) return 'PB0007';

  const title = input.focusTitle;
  const forgot = (input.patternForgotCount ?? 0) >= 3;
  if (forgot) {
    if (looksBrush(title)) return 'PB0001';
    if (looksPack(title)) return 'PB0002';
    if (looksSleep(title)) return 'PB0003';
    if (looksRead(title)) return 'PB0004';
    return 'PB0008';
  }

  if (input.skipReasonForgot) {
    if (looksBrush(title)) return 'PB0001';
    if (looksPack(title)) return 'PB0002';
    return 'PB0017';
  }

  const code = (input.proposalCode ?? '').trim();
  if (
    code === PROPOSAL.suggest_move_after_dinner ||
    code === PROPOSAL.suggest_move_after_school
  ) {
    if (looksBrush(title)) return 'PB0001';
    if (looksPack(title)) return 'PB0002';
    return 'PB0008';
  }
  if (code === PROPOSAL.support_overdue) return 'PB0006';
  if (code === PROPOSAL.open_today) {
    if (looksBrush(title)) return 'PB0001';
    if (looksPack(title)) return 'PB0002';
  }

  if (looksBrush(title)) return 'PB0001';
  if (looksPack(title)) return 'PB0002';
  if (looksSleep(title)) return 'PB0003';
  if (looksRead(title)) return 'PB0004';

  if (input.blueprintSparse) return 'PB0020';
  return null;
}
